mat64/solve.go

```diff
@@ -0,0 +1,87 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat64
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/lapack/lapack64"
+)
+
+// Solve solves a minimum-norm solution to a system of linear equations defined
+// by the matrices a and b. If a is singular or near-singular a Condition error
+// is returned. Please see the documentation for Condition for more information.
+//
+// The minimization problem solved depends on the input parameters.
+//  1. If m >= n and trans == false, find x such that || a*x - b||_2 is minimized.
+//  2. If m < n and trans == false, find the minimum norm solution of a * x = b.
+//  3. If m >= n and trans == true, find the minimum norm solution of a^T * x = b.
+//  4. If m < n and trans == true, find X such that || a*x - b||_2 is minimized.
+// The solution matrix, x, is stored in place into the receiver.
+func (m *Dense) Solve(a, b Matrix) error {
+	ar, ac := a.Dims()
+	br, bc := b.Dims()
+	if ar != br {
+		panic(ErrShape)
+	}
+
+	aMat, _ := untranspose(a)
+	bMat, _ := untranspose(b)
+	m.reuseAs(ac, bc)
+	var restore func()
+	if m == aMat {
```

Comment:
What would be fast pathed? In LU b still has to be copied into m. In QR it still has to be copied twice because A and B are modified independently in Gels.